File: algorithm/stack4.py
# https://www.acmicpc.net/problem/9012
# 실버4

n = int(input())

for i in range(n):
    ps=input()
    stack=[]
    end=0
    for char in ps:
        if char == '(':
            stack.append(char)
        else:
            if len(stack) == 0:
                print('NO')
                end=1
                break
            else:
                stack.pop()
    if end == 0:
        if len(stack) == 0:
            print('YES')
        else:
            print('NO')
    

File: algorithm-js/stack4.js
// 실버4 https://www.acmicpc.net/problem/9012
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

const [N, ...inputStrings] = input;

inputStrings.forEach(inputString => {
    const stack = [];
    let end = false;

    for (let i = 0; i < inputString.length; i++) {
        if (inputString[i] === '(') {
            stack.push(inputString[i]);
        } else {
            if (stack.length === 0) {
                end = true;
                console.log('NO');
                break;
            } else {
                stack.pop();
            }
        }

    }

    if (!end) {
        if (stack.length === 0) {
            console.log('YES');
        } else {
            console.log('NO');
        }
    }
});
